packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1970 @@
+   * @param requestedPermissions - The requested permissions.
+   */
+  private validateRequestedPermissions(
+    origin: OriginString,
+    requestedPermissions: RequestedPermissions | unknown,
+  ): void {
+    if (!isPlainObject(requestedPermissions)) {
+      throw invalidParams({
+        message: `Requested permissions for origin "${origin}" is not a plain object.`,
+        data: { origin, requestedPermissions },
+      });
+    }
+
+    if (Object.keys(requestedPermissions).length === 0) {
+      throw invalidParams({
+        message: `Permissions request for origin "${origin}" contains no permissions.`,
+        data: { requestedPermissions },
+      });
+    }
+
+    for (const methodName of Object.keys(requestedPermissions)) {
+      const permission = requestedPermissions[methodName];
+      if (
+        !isPlainObject(permission) ||
+        (permission.parentCapability !== undefined &&
+          methodName !== permission.parentCapability)
+      ) {
+        throw invalidParams({
+          message: `Permissions request for origin "${origin}" contains invalid requested permission(s).`,
+          data: { origin, requestedPermissions },
+        });
+      }
+
+      if (!this.getTargetKey(methodName)) {
+        throw methodNotFound({
+          method: methodName,
+          data: { origin, requestedPermissions },
+        });
+      }
+    }
+  }
+
+  /**
+   * Adds a request to the {@link ApprovalController} using the
+   * {@link AddApprovalRequest} action. Also validates the resulting approved
+   * permissions request, and throws an error if validation fails.
+   *
+   * @param permissionsRequest - The permissions request object.
+   * @returns The approved permissions request object.
+   */
+  private async requestUserApproval(permissionsRequest: PermissionsRequest) {
+    const { origin, id } = permissionsRequest.metadata;
+    const approvedRequest = await this.messagingSystem.call(
+      'ApprovalController:addRequest',
+      {
+        id,
+        origin,
+        requestData: permissionsRequest,
+        type: MethodNames.requestPermissions,
+      },
+      true,
+    );
+
+    this.validateApprovedPermissions(approvedRequest, { id, origin });
+    return approvedRequest as PermissionsRequest;
+  }
+
+  /**
+   * Validates an approved {@link PermissionsRequest} object. The approved
+   * request must have the required `metadata` and `permissions` properties,
+   * the `id` and `origin` of the `metadata` must match the original request
+   * metadata, and the requested permissions must be valid per
+   * {@link PermissionController.validateRequestedPermissions}. Any extra
+   * metadata properties are ignored.
+   *
+   * An error is thrown if validation fails.
+   *
+   * @param approvedRequest - The approved permissions request object.
+   * @param originalMetadata - The original request metadata.
+   */
+  private validateApprovedPermissions(
+    approvedRequest: PermissionsRequest | unknown,
```

Comment:
Nit: Any type union including `unknown` is equivalent to `unknown`

```suggestion
    approvedRequest: unknown,
```

Reply:
Actually I've just noticed this being done in a few other places too, mostly in the tests I think. I searched for `| unknown`. I think all of those type unions can be reduced to `unknown`.

Reply:
91cf3df